query/src/sql/planner/binder/sort.rs

```diff
@@ -46,91 +47,114 @@ pub struct OrderItem<'a> {
 }
 
 impl<'a> Binder {
-    pub(super) fn analyze_order_items(
+    pub(super) async fn analyze_order_items(
         &mut self,
         from_context: &BindContext,
-        scalar_items: &HashMap<IndexType, ScalarItem>,
+        scalar_items: &mut HashMap<IndexType, ScalarItem>,
         projections: &[ColumnBinding],
         order_by: &'a [OrderByExpr<'a>],
         distinct: bool,
     ) -> Result<OrderItems<'a>> {
         let mut order_items = Vec::with_capacity(order_by.len());
         for order in order_by {
-            if let Expr::ColumnRef {
-                database: ref database_name,
-                table: ref table_name,
-                column: ref ident,
-                ..
-            } = order.expr
-            {
-                // We first search the identifier in select list
-                let mut found = false;
-                for item in projections.iter() {
-                    if item.column_name == ident.name {
-                        order_items.push(OrderItem {
-                            expr: order.clone(),
-                            index: item.index,
-                            name: item.column_name.clone(),
-                            need_project: scalar_items.get(&item.index).map_or(
-                                false,
-                                |scalar_item| {
-                                    !matches!(&scalar_item.scalar, Scalar::BoundColumnRef(_))
-                                },
-                            ),
-                        });
-                        found = true;
-                        break;
+            match &order.expr {
+                Expr::ColumnRef {
+                    database: ref database_name,
+                    table: ref table_name,
+                    column: ref ident,
+                    ..
+                } => {
+                    // We first search the identifier in select list
+                    let mut found = false;
+                    for item in projections.iter() {
+                        if item.column_name == ident.name {
+                            order_items.push(OrderItem {
+                                expr: order.clone(),
+                                index: item.index,
+                                name: item.column_name.clone(),
+                                need_eval_scalar: scalar_items.get(&item.index).map_or(
+                                    false,
+                                    |scalar_item| {
+                                        !matches!(&scalar_item.scalar, Scalar::BoundColumnRef(_))
+                                    },
+                                ),
+                            });
+                            found = true;
+                            break;
+                        }
                     }
-                }
 
-                if found {
-                    continue;
-                }
+                    if found {
```

Comment:
if `need_eval_scalar` is true, should we call `scalar_items.insert` here?

Reply:
No, because the matching branch is `Expr::ColumnRef`